tests/stratagus/test_missile_fire.cpp

```diff
@@ -0,0 +1,172 @@
+		BurningBuildingFrames.resize(size);
+		for (int i = 0; i < size; ++i) {
+			// reverse order as in CclDefineBurningBuilding
+			BurningBuildingFrames[size - 1 - i].Percent = frames[i].first;
+			BurningBuildingFrames[size - 1 - i].Missile = frames[i].second;
+		}
+	}
+
+} // anonymous namespace
+
+TEST_CASE("missile_fire")
+{
+	BurningBuildingFrames.clear();
+
+	SUBCASE("size=1") {
+		setFrames({
+			{  0, M(10)},
+		});
+
+		CHECK(MissileBurningBuilding(  0) == M(10));
+		CHECK(MissileBurningBuilding(  1) == M(10));
+		CHECK(MissileBurningBuilding(100) == M(10));
+	}
+
+	SUBCASE("large steps") {
+		setFrames({
+			{  0, M(10)},
+			{ 50, M(11)},
+		});
+
+		SUBCASE("normal") {
+			CHECK(MissileBurningBuilding(  0) == M(10));
+			CHECK(MissileBurningBuilding(  1) == M(10));
+			CHECK(MissileBurningBuilding( 49) == M(10));
+			CHECK(MissileBurningBuilding( 50) == M(11));
+			CHECK(MissileBurningBuilding( 51) == M(11));
+			CHECK(MissileBurningBuilding(100) == M(11));
+		}
+
+		SUBCASE("out of bounds") {
+			CHECK(MissileBurningBuilding( -1) == nullptr);
+			CHECK(MissileBurningBuilding(101) == M(11));
+		}
+	}
+
+	SUBCASE("partially dense")
+	{
+		setFrames({
+			{  0, M(10)},
+			{  1, M(11)},
+			{  2, M(12)},
+			{ 65, M(13)},
+			{ 66, M(14)},
+			{ 67, M(15)},
+			{ 99, M(16)},
+			{100, M(17)},
+		});
+
+		CHECK(MissileBurningBuilding(  0) == M(10));
+		CHECK(MissileBurningBuilding(  1) == M(11));
+		CHECK(MissileBurningBuilding(  2) == M(12));
+		CHECK(MissileBurningBuilding(  3) == M(12));
+
+		CHECK(MissileBurningBuilding( 50) == M(12));
+
+		CHECK(MissileBurningBuilding( 64) == M(12));
+		CHECK(MissileBurningBuilding( 65) == M(13));
+		CHECK(MissileBurningBuilding( 66) == M(14));
+		CHECK(MissileBurningBuilding( 67) == M(15));
+		CHECK(MissileBurningBuilding( 68) == M(15));
+
+		CHECK(MissileBurningBuilding( 98) == M(15));
+		CHECK(MissileBurningBuilding( 99) == M(16));
+		CHECK(MissileBurningBuilding(100) == M(17));
+	}
+
+	SUBCASE("repeated values and null")
+	{
+		setFrames({
+			{  0, M(10)},
+			{ 80, nullptr},
+			{ 90, M(10)},
+		});
+
+		CHECK(MissileBurningBuilding( 79) == M(10));
+		CHECK(MissileBurningBuilding( 80) == nullptr);
+		CHECK(MissileBurningBuilding( 89) == nullptr);
+		CHECK(MissileBurningBuilding( 90) == M(10));
+	}
+
+	SUBCASE("first key larger") {
+		setFrames({
+			{ 10, M(1000)},
+		});
+
+		CHECK(MissileBurningBuilding( 9) == nullptr);
+		CHECK(MissileBurningBuilding(10) == M(1000));
+	}
+
+	SUBCASE("duplicate keys")
+	{
+		setFrames({
+			{  0, M(10)},
+			{ 20, M(11)},
+			{ 20, M(12)},
+			{ 60, M(13)},
+		});
+
+		CHECK(MissileBurningBuilding(19) == M(10));
+		CHECK(MissileBurningBuilding(20) == M(12));
+		CHECK(MissileBurningBuilding(21) == M(12));
+	}
+
+	SUBCASE("wrong key order") {
+		setFrames({
+			{ 75, M(12)},
+			{ 50, M(11)},
+			{  0, M(10)},
+		});
+
+		CHECK(MissileBurningBuilding(  0) == M(10));
+		CHECK(MissileBurningBuilding( 20) == M(10));
+		CHECK(MissileBurningBuilding( 50) == M(10));
```

Comment:
I am not sure if the test of bad behaviour should be kept.

Reply:
We should probably have a check in `CclDefineBurningBuilding` that order is correct, and no null value.
And so we can test this checking function instead.

Reply:
Added and used function is_BurningBuildingFramesSorted.
Null value is allowed to mark not-burning building.